Core/GDCore/IDE/Events/EventsVariableReplacer.cpp

```diff
@@ -260,29 +300,97 @@ class GD_CORE_API ExpressionVariableReplacer
+      return false;
+    }
+    const auto &oldToNewVariableNames =
+        currentVariablesRenamingChangesetNode->oldToNewVariableNames;
     if (oldToNewVariableNames.count(variableName) >= 1) {
       variableName = oldToNewVariableNames.find(variableName)->second;
       hasDoneRenaming = true;
       return true;
-    } else if (removedVariableNames.count(variableName) >= 1) {
+    } else if (
+      // Only the root variable is checked for removing.
+      currentVariablesRenamingChangesetNode ==
+                   &variablesRenamingChangesetRoot &&
+               removedVariableNames.count(variableName) >= 1) {
       removedVariableUsed = true;
       return true;
     }
 
     return false;  // Nothing was changed or done.
   }
 
+  void RenameVariableAndVisitChild(gd::String &variableName,
+                                   ExpressionNode *childNode) {
+    // `variableName` is modified by
+    // `RenameOrRemoveVariableOfTargetVariableContainer`.
+    const gd::String oldVariableName = variableName;
+    RenameOrRemoveVariableOfTargetVariableContainer(variableName);
+    if (childNode) {
+      bool hasBeenPushed =
+          PushVariablesRenamingChangesetNodeForVariable(oldVariableName);
+      childNode->Visit(*this);
+      PopVariablesRenamingChangesetNode(hasBeenPushed);
+    }
+  }
+
+  void PushVariablesRenamingChangesetRoot() {
+    variablesRenamingChangesetNodeStack.push_back(&variablesRenamingChangesetRoot);
+  }
+
+  void PushVariablesRenamingChangesetNodeForIgnoredVariables() {
+    variablesRenamingChangesetNodeStack.push_back(nullptr);
+  }
+
+  const gd::VariablesRenamingChangesetNode *GetCurrentVariablesRenamingChangesetNode() {
+    return variablesRenamingChangesetNodeStack.size() == 0 ?
+    &variablesRenamingChangesetRoot :
+        variablesRenamingChangesetNodeStack
+            [variablesRenamingChangesetNodeStack.size() - 1];
+  }
+
+  bool PushVariablesRenamingChangesetNodeForVariable(const gd::String& variableName) {
+    const auto *currentVariablesRenamingChangesetNode = GetCurrentVariablesRenamingChangesetNode();
+    if (!currentVariablesRenamingChangesetNode) {
+      // There were already no more change on a parent.
+      return false;
+    }
+    const auto &childVariablesRenamingChangesetNodeItr =
+        currentVariablesRenamingChangesetNode->modifiedVariables.find(
+            variableName);
+    if (childVariablesRenamingChangesetNodeItr ==
+        currentVariablesRenamingChangesetNode->modifiedVariables.end()) {
+      // There is no more change on the current variable child.
+      variablesRenamingChangesetNodeStack.push_back(nullptr);
+    }
+    else {
+      variablesRenamingChangesetNodeStack.push_back(
+          childVariablesRenamingChangesetNodeItr->second.get());
+    }
+    return true;
+  }
+
+  void PopVariablesRenamingChangesetNode(bool hasBeenPushed) {
+    if (hasBeenPushed) {
+      variablesRenamingChangesetNodeStack.push_back(&variablesRenamingChangesetRoot);
```

Comment:
Can you comment why a `pop`  operation is doing a `push_back`?

Reply:
A bad copy-paste I guess, but it's odd that tests passed with it because I tested several renaming on same level. Maybe the test wasn't deep enough.